### PR TITLE
Add/remove User Capabilities Tests

### PR DESCRIPTION
**Pull request by aditipande96**
Signed-off-by: Aditi Pande <aditi.pande@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [ ] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by aditipande96** [comment]
Test results:
PASSED tests/csm/rest/test_iam_users.py::TestIamUserRGW::test_39026
PASSED tests/csm/rest/test_iam_users.py::TestIamUserRGW::test_39027
PASSED tests/csm/rest/test_iam_users.py::TestIamUserRGW::test_39028

Collection logs:
ERROR tests/security/test_cortx_port_scanner_kubectl_svc.py
ERROR tests/security/test_cortx_port_scanner_netstat.py
================= 2032 tests collected, 2 errors in 9.68s ====================

File: tests/csm/rest/test_iam_users.py [hunk 1]
```diff
@@ -2126,3 +2126,207 @@ def test_39217(self):
...
+        self.log.info(
+            "Step 1: Login using csm user and create a user with some capabilities")
+        payload = self.csm_obj.iam_user_payload_rgw(user_type="valid")
+        user_cap = "usage=read;buckets=write;users=read"
+        payload.update({"user_caps": user_cap})
+        self.log.info("payload :  %s", payload)
+        resp1 = self.csm_obj.create_iam_user_rgw(payload)
+        assert resp1.status_code == HTTPStatus.CREATED, \
+            "User could not be created"
+        uid = resp1.json()['tenant'] + "$" + payload["uid"]
```
Comment (swapnil-khandare) [inline tests/csm/rest/test_iam_users.py:2305]: uid is not added to created_iam_users

File: tests/csm/rest/test_iam_users.py [hunk 2]
```diff
@@ -2126,3 +2126,207 @@ def test_39217(self):
...
+        payload = {}
+        payload.update({"user_caps": user_cap})
+        resp = self.csm_obj.add_user_caps_rgw(uid, payload)
+        self.log.info("Verify Response : %s", resp)
+        assert_utils.assert_true(resp.status_code == HTTPStatus.BAD_REQUEST,
+                                     "Invalid caps added")
+        get_resp = self.csm_obj.get_iam_user(user=uid)
+        assert_utils.assert_true(get_resp.status_code == HTTPStatus.OK, "Get IAM user failed")
+        self.log.info("STEP 3: Verify added capabilities")
+        diff_items = self.csm_obj.verify_caps(user_cap, get_resp.json()["caps"])
```
Comment (swapnil-khandare) [inline tests/csm/rest/test_iam_users.py:2317]: here user_cap should be initial caps, but user_cap is getting updated on line 2307.

File: tests/csm/rest/test_iam_users.py [hunk 3]
```diff
@@ -2126,3 +2126,207 @@ def test_39217(self):
...
+        payload = {}
+        payload.update({"user_caps": user_cap})
+        resp = self.csm_obj.remove_user_caps_rgw(uid, payload)
+        self.log.info("Verify Response : %s", resp)
+        assert_utils.assert_true(resp.status_code == HTTPStatus.BAD_REQUEST,
+                                     "Attempt to remove invalid caps")
+        get_resp = self.csm_obj.get_iam_user(user=uid)
+        assert_utils.assert_true(get_resp.status_code == HTTPStatus.OK, "Get IAM user failed")
+        self.log.info("STEP 5: Verify removed capabilities")
+        assert_utils.assert_false(len(get_resp.json()["caps"]) == 0,
```
Comment (swapnil-khandare) [inline tests/csm/rest/test_iam_users.py:2330]: cap should not be zero. It has 3 initial caps

File: tests/csm/rest/test_iam_users.py (hunk 1 above)
Comment (aditipande96) [reply tests/csm/rest/test_iam_users.py:2305]: will add it

Comment (aditipande96) [reply tests/csm/rest/test_iam_users.py:2305]: added

File: tests/csm/rest/test_iam_users.py (hunk 2 above)
Comment (aditipande96) [reply tests/csm/rest/test_iam_users.py]: changed verification code

File: tests/csm/rest/test_iam_users.py (hunk 3 above)
Comment (aditipande96) [reply tests/csm/rest/test_iam_users.py]: changed verification code.

**Review by swapnil-khandare — APPROVED** [review]

File: tests/csm/rest/test_iam_users.py [hunk 4]
```diff
@@ -2126,3 +2126,207 @@ def test_39217(self):
...
+        test_case_name = cortxlogging.get_frame()
+        self.log.info("##### Test started -  %s #####", test_case_name)
+        self.log.info(
+            "Step 1: Login using csm user and create a user with some capabilities")
+        payload = self.csm_obj.iam_user_payload_rgw(user_type="valid")
+        user_cap = "usage=read;buckets=read;users=read"
+        payload.update({"user_caps": user_cap})
+        self.log.info("payload :  %s", payload)
+        resp = self.csm_obj.create_iam_user_rgw(payload)
+        assert resp.status_code == HTTPStatus.CREATED, \
```
Comment (RajeshPratapSInghChouhan) [inline tests/csm/rest/test_iam_users.py:2147]: please use assert_utils if possible

File: tests/csm/rest/test_iam_users.py [hunk 5]
```diff
@@ -2126,3 +2126,207 @@ def test_39217(self):
...
+        test_case_name = cortxlogging.get_frame()
+        self.log.info("##### Test started -  %s #####", test_case_name)
+        self.log.info(
+            "Step 1: Login using csm user and create a user with some capabilities")
+        payload = self.csm_obj.iam_user_payload_rgw(user_type="valid")
+        user_cap = "buckets=read,write"
+        payload.update({"user_caps": user_cap})
+        self.log.info("payload :  %s", payload)
+        resp1 = self.csm_obj.create_iam_user_rgw(payload)
+        assert resp1.status_code == HTTPStatus.CREATED, \
```
Comment (RajeshPratapSInghChouhan) [inline tests/csm/rest/test_iam_users.py:2199]: please use assert_utils if possible

File: tests/csm/rest/test_iam_users.py [hunk 6]
```diff
@@ -2126,3 +2126,207 @@ def test_39217(self):
...
+        test_case_name = cortxlogging.get_frame()
+        self.log.info("##### Test started -  %s #####", test_case_name)
+        self.log.info(
+            "Step 1: Login using csm user and create a user with some capabilities")
+        payload = self.csm_obj.iam_user_payload_rgw(user_type="valid")
+        user_cap = "usage=read;buckets=write;users=read"
+        payload.update({"user_caps": user_cap})
+        self.log.info("payload :  %s", payload)
+        resp1 = self.csm_obj.create_iam_user_rgw(payload)
+        assert resp1.status_code == HTTPStatus.CREATED, \
```
Comment (RajeshPratapSInghChouhan) [inline tests/csm/rest/test_iam_users.py:2303]: please use assert_utils if possible

**Review by RajeshPratapSInghChouhan — APPROVED** [review]

File: tests/csm/rest/test_iam_users.py [hunk 7]
```diff
@@ -2126,3 +2126,207 @@ def test_39217(self):
...
+        assert_utils.assert_true(get_resp.status_code == HTTPStatus.OK, "Get IAM user failed")
+        self.log.info("STEP 3: Verify added capabilities")
+        assert_utils.assert_true(get_resp.json()["caps"] == resp1.json()["caps"],
+                                 "Invalid capabilities added")
+        self.log.info("Step 4: Remove invalid capabilities")
+        payload = {}
+        payload.update({"user_caps": user_cap})
+        resp = self.csm_obj.remove_user_caps_rgw(uid, payload)
+        self.log.info("Verify Response : %s", resp)
+        assert_utils.assert_true(resp.status_code == HTTPStatus.BAD_REQUEST,
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_iam_users.py:2325]: Can  u check response message here.

File: tests/csm/rest/test_iam_users.py (hunk 5 above)
Comment (kachhwahadivya) [reply tests/csm/rest/test_iam_users.py:2199]: For debugging assert are easier to debug than assert_utils, as It logs failing tests where with assert utils it logs assert_util file.
Any specific reason for promoting this?

Comment (RajeshPratapSInghChouhan) [reply tests/csm/rest/test_iam_users.py:2199]: > Any specific reason for promoting this?

for uniformity i think.

but if raw asserts helps us to get more info for some specific case,
like a REST API response, should be ok.

File: tests/csm/rest/test_iam_users.py [hunk 8]
```diff
@@ -2126,3 +2126,207 @@ def test_39217(self):
...
+        self.log.info("Create bucket successful for user")
+        test_file = "test-object.txt"
+        file_path_upload = os.path.join(TEST_DATA_FOLDER, test_file)
+        if os.path.exists(file_path_upload):
+            os.remove(file_path_upload)
+        if not os.path.isdir(TEST_DATA_FOLDER):
+            self.log.debug("File path not exists, create a directory")
+            system_utils.execute_cmd(cmd=common_cmd.CMD_MKDIR.format(TEST_DATA_FOLDER))
+        system_utils.create_file(file_path_upload, self.file_size)
+        self.log.info("Step: Verify put object.")
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_iam_users.py:2166]: changed

File: tests/csm/rest/test_iam_users.py (hunk 7 above)
Comment (aditipande96) [reply tests/csm/rest/test_iam_users.py]: done

File: tests/csm/rest/test_iam_users.py (hunk 5 above)
Comment (RajeshPratapSInghChouhan) [reply tests/csm/rest/test_iam_users.py:2199]: @ketanarlulkar created : https://github.com/Seagate/cortx-test/issues/1392

**Review by kachhwahadivya — APPROVED** [review]

**Review by ketanarlulkar — APPROVED** [review]

**Pull request merged**